Dependencies/xr/Source/ARCore/XR.cpp

```diff
@@ -642,6 +655,158 @@ namespace xr
+                    switch (arAvailability)
+                    {
+                        case AR_AVAILABILITY_SUPPORTED_APK_TOO_OLD:
+                        case AR_AVAILABILITY_SUPPORTED_INSTALLED:
+                        case AR_AVAILABILITY_SUPPORTED_NOT_INSTALLED:
+                            tcs.complete(true);
+                            break;
+                        case AR_AVAILABILITY_UNKNOWN_CHECKING:
+                            std::this_thread::sleep_for(std::chrono::milliseconds(200));
+                            break;
+                        default:
+                            tcs.complete(false);
+                            break;
+                    }
+
+                    if (tcs.completed())
+                    {
+                        break;
+                    }
+                }
+
+                if (!tcs.completed())
+                {
+                    tcs.complete(false);
+                }
+            });
+
+            return tcs.as_task();
+        }
+
+        // VR and inline sessions are not supported at this time.
+        return arcana::task_from_result<std::exception_ptr>(false);
+    }
+
+    bool CheckARCoreInstallStatus(bool requestInstall)
+    {
+        ArInstallStatus install_status;
+        ArStatus installStatus = ArCoreApk_requestInstall(
+                GetEnvForCurrentThread(), GetMainActivity(), requestInstall, &install_status);
+        return installStatus == AR_SUCCESS && install_status == AR_INSTALL_STATUS_INSTALLED;
+    }
+
+    arcana::task<void, std::exception_ptr> CheckAndInstallARCoreAsync(const Babylon::JsRuntimeScheduler& runtimeScheduler)
+    {
+        auto task = arcana::task_from_result<std::exception_ptr>();
+
+        // Check if ARCore is already installed.
+        if (!CheckARCoreInstallStatus(false))
+        {
+            arcana::task_completion_source<void, std::exception_ptr> installTcs{};
+
+            // Add a resume callback, which will check if ARCore has been successfully installed upon app resume.
+            auto resumeTicket{AddResumeCallback([installTcs]() mutable {
+                if (!CheckARCoreInstallStatus(false))
+                {
+                    // ARCore not installed, throw an error.
+                    std::ostringstream message;
+                    message << "ARCore not installed.";
+                    installTcs.complete(arcana::make_unexpected(make_exception_ptr(std::runtime_error{message.str()})));
+                }
+                else
+                {
+                    // ARCore installed successfully, complete the promise.
+                    installTcs.complete();
+                }
+            })};
+
+            // Kick off the install request, and set the task for our caller to wait on.
+            CheckARCoreInstallStatus(true);
+            task = installTcs.as_task().then(runtimeScheduler, arcana::cancellation::none(), [resumeTicket = std::move(resumeTicket)](){});
```

Comment:
All this continuation does is destroy the ticket, and the underlying ticketed_collection is already thread safe, so you should be able to just use `inline_scheduler` here.